### PR TITLE
Add logic to roll forward to patch versions of .NET Core for self-contained apps but not shared framework apps or libraries

### LINKED ISSUE #983: Reconsider implicitly using the latest patch version of the targeted version of .NET Core in the SDK
From @eerhardt:

> The issue is someone upgrades their SDK on their developer machine, which brings in the latest patch version (1.1.1 in this case) and then publishes their app to their host machine. The host machines aren't updated to 1.1.1, because it just got released and these things take time. Now the app breaks on a host machine because 1.1.1 isn't installed.

See more comments/discussion on #860.

### PR DESCRIPTION
**Pull request by dsplaisted**
Fixes #983

### CODE REVIEW

**Comment by dasMulli** [comment]
Am I correct in assuming the change will affect the version of the NuGet package being referenced? This would then require a restore using the same properties when called through msbuild directly.. (breaking change?).
Are there any downsides to just downgrading the version written to the runtimeconfig.json to a `.0` patch and using the latest NuGet?

**Comment by dsplaisted** [comment]
By itself this doesn't change anything, as we only want to apply the new logic to patches that we ship in the future.

Previously we were rolling forward to the latest patch version (that existed at the time the SDK ships) for all apps.  This changes the behavior so that in the future, we will only roll forward to patch versions for self-contained apps, while shared framework apps will continue to target version x.y.0.

There was some discussion about referencing newer NuGet packages but downgrading the version written to the runtimeconfig.json.  However, that feels a bit dangerous to me.  In general we don't support compiling against a later version of something and running on a previous version of it.  It's *likely* that we won't add any APIs in patch versions or make any other changes that would break this, but I think it's simpler and safer to keep them in sync.

Does that help explain things?

**Comment by dasMulli** [comment]
Yes, thanks!
So it comes down to "protect against APIs that won't be available" vs "make sure fixes in the ref package are consumed" 😂 

So this likely requires `/t:Restore;Publish /p:[PropsForSCD]` when publishing with msbuild or it might use outdated packages silently?

**Comment by dsplaisted** [comment]
@dotnet-bot test Windows_NT_FullFramework Debug

**Comment by dsplaisted** [comment]
Yeah, if we have another issue where there's an issue we want to fix with the package but not the runtime, then we might need to think about a split between the package version and the version that goes in runtimeconfig.json.  I hope it doesn't come to that :-)

> So this likely requires `/t:Restore;Publish /p:[PropsForSCD]` when publishing with msbuild or it might use outdated packages silently?

- Combining the `Restore` target with pretty much any other target in the same msbuild invocation isn't supported.  .props and .targets files that are written by the restore target won't be imported when it runs subsequent targets.
- The properties added in this PR are implementation details and test hooks.  You shouldn't need to set them.  If you want to target a specific patch, then you do so by setting the `RuntimeFrameworkVersion` property (and this PR doesn't change that).

**Comment by dsplaisted** [comment]
@dotnet-bot test Windows_NT_FullFramework Debug

**Comment by dsplaisted** [comment]
@MattGertz for approval

**Comment by MattGertz** [comment]
@dsplaisted Template?

**Review by nguerrera — APPROVED** [review]

File: src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.targets [hunk 1]
```diff
@@ -73,33 +73,72 @@ Copyright (c) .NET Foundation. All rights reserved.
...
+        <ImplicitRuntimeFrameworkVersionForSelfContainedApp>$(BundledNETCoreAppPackageVersion)</ImplicitRuntimeFrameworkVersionForSelfContainedApp>
+      </PropertyGroup>
+    </When>
     
-    <!-- Default to use the version of the framework runtime matching the target framework version-->
-    <RuntimeFrameworkVersion Condition="'$(RuntimeFrameworkVersion)' == ''">$(_TargetFrameworkVersionWithoutV)</RuntimeFrameworkVersion>
+    <!-- If not covered by the previous cases, use the target framework version for the implicit RuntimeFrameworkVersions -->
+    <Otherwise>
+      <PropertyGroup>
+        <ImplicitRuntimeFrameworkVersionForSharedApp>$(_TargetFrameworkVersionWithoutV)</ImplicitRuntimeFrameworkVersionForSharedApp>
```
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/build/Microsoft.NET.Sdk.DefaultItems.targets]: I'd use the term `FrameworkDependent` in code.  This is the term we use in documentation:  [https://docs.microsoft.com/en-us/dotnet/articles/core/deploying/](https://docs.microsoft.com/en-us/dotnet/articles/core/deploying/)

And also in the rest of the SDK code: [https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/ProjectContext.cs#L19-L23](https://github.com/dotnet/sdk/blob/master/src/Tasks/Microsoft.NET.Build.Tasks/ProjectContext.cs#L19-L23)

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildANetCoreApp.cs [hunk 2]
```diff
@@ -30,37 +30,75 @@ public class GivenThatWeWantToBuildANetCoreApp : SdkTest
...
+        {
+            string testIdentifier = "Framework_targeting_" + (isExe ? "App_" : "Lib_") + (selfContained ? "SelfContained" : "SharedFramework");
+
+            It_targets_the_right_framework(testIdentifier, "netcoreapp1.1", null, selfContained, isExe, expectedFrameworkVersion, expectedFrameworkVersion,
+                "/p:ImplicitRuntimeFrameworkVersionForSharedNetCoreApp1_1=1.1.1");
+        }
+
+        private void It_targets_the_right_framework(
+            string testIdentifier,
+            string targetFramework, string runtimeFrameworkVersion,
```
Comment (eerhardt) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildANetCoreApp.cs]: NIT: I think putting some args on new lines, and some inline makes this hard to read.

File: test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildANetCoreApp.cs [hunk 3]
```diff
@@ -30,37 +30,75 @@ public class GivenThatWeWantToBuildANetCoreApp : SdkTest
         [InlineData("netcoreapp1.1", null, "1.1.2", "1.1.2")]
         [InlineData("netcoreapp1.1", "1.1.0", "1.1.0", "1.1.0")]
         [InlineData("netcoreapp1.1.1", null, "1.1.1", "1.1.1")]
-        public void It_targets_the_right_shared_framework(string targetFramework, string runtimeFrameworkVersion,
+        private void It_targets_the_right_shared_framework(string targetFramework, string runtimeFrameworkVersion,
```
Comment (eerhardt) [inline test/Microsoft.NET.Build.Tests/GivenThatWeWantToBuildANetCoreApp.cs]: Why did this change to `private`?

**Review by eerhardt — APPROVED** [review]

**Comment by dsplaisted** [comment]
@MattGertz for approval

**Customer scenario**

Targeting and running an app.  This change means we will default to targeting the latest patch version of .NET Core for self-contained apps (so they get the latest updates), but for framework-dependent apps, we will target the `Major.Minor.0` version, and rely on the runtime roll-forward policy.  This means framework-dependent apps won't fail to run when deployed to hosts that haven't updated the shared framework.

**Bugs this fixes:** 

#983

**Workarounds, if any**

Explicitly set `RuntimeFrameworkVersion`.

**Risk**

Low.

**Performance impact**

Low - this is an update to logic which is statically evaluated by MSBuild

**Is this a regression from a previous update?**

Yes.  We had policy (first introduced in #878) which would cause regressions in SDK updates.

**Root cause analysis:**

Hosts that provide a shared framework aren't necessarily immediately updated to a new runtime when we release that runtime together with an SDK.

**How was the bug found?**

Hosters and developers deploying to hosts reported the issue when they found that apps were broken when publishing to the hosts.

**Pull request merged**